File: src/Core/Template.fs
namespace ARCtrl

open Fable.Core
open ARCtrl.Helper

[<AttachMembers>]
type Organisation =
| [<CompiledName "DataPLANT">] DataPLANT
| Other of string
    static member ofString (str:string) = 
        match str.ToLower() with
        | "dataplant" -> DataPLANT
        | _ -> Other str

    override this.ToString() =
        match this with
        | DataPLANT -> "DataPLANT"
        | Other anyElse -> anyElse

    member this.IsOfficial() = this = DataPLANT

[<AttachMembers>]
type Template(id: System.Guid, table: ArcTable, ?name: string, ?description, ?organisation: Organisation, ?version: string, ?authors: ResizeArray<Person>, 
    ?repos: ResizeArray<OntologyAnnotation>, ?tags: ResizeArray<OntologyAnnotation>, ?lastUpdated: System.DateTime) =

    let name = defaultArg name ""
    let description = defaultArg description ""
    let organisation = defaultArg organisation (Other "Custom Organisation")
    let version = defaultArg version "0.0.0"
    let authors = defaultArg authors <| ResizeArray()
    let repos = defaultArg repos <| ResizeArray()
    let tags = defaultArg tags <| ResizeArray()
    let lastUpdated = defaultArg lastUpdated (System.DateTime.Now)

    member val Id : System.Guid = id with get, set
    member val Table : ArcTable = table with get, set
    member val Name : string = name with get, set
    member val Description : string = description with get, set
    member val Organisation : Organisation = organisation with get, set
    member val Version : string = version with get, set
    member val Authors : ResizeArray<Person> = authors with get, set
    member val EndpointRepositories : ResizeArray<OntologyAnnotation> = repos with get, set
    member val Tags : ResizeArray<OntologyAnnotation> = tags with get, set
    member val LastUpdated : System.DateTime = lastUpdated with get, set

    static member make id table name description organisation version authors repos tags lastUpdated =
        Template(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated)

    static member create(id, table, ?name, ?description, ?organisation, ?version, ?authors, ?repos, ?tags, ?lastUpdated) =
        Template(id, table, ?name=name, ?description = description, ?organisation=organisation, ?version=version, ?authors=authors, ?repos=repos, ?tags=tags, ?lastUpdated=lastUpdated)

    static member init(templateName: string) =
        let guid = System.Guid.NewGuid()
        let table = ArcTable.init(templateName)
        Template(guid, table, templateName)

    member this.SemVer
        with get() = ARCtrl.Helper.SemVer.SemVer.tryOfString this.Version

    /// <summary>
    /// Use this function to check if this Template and the input Template refer to the same object.
    ///
    /// If true, updating one will update the other due to mutability.
    /// </summary>
    /// <param name="other">The other Template to test for reference.</param>
    member this.ReferenceEquals (other: Template) = System.Object.ReferenceEquals(this,other)

    member this.StructurallyEquals (other: Template) =
        this.GetHashCode() = other.GetHashCode()

    // custom check
    override this.Equals other =
        match other with
        | :? Template as template ->
            this.StructurallyEquals(template)
        | _ -> 
            false

    override this.GetHashCode() =
        [|
            box (this.Id.ToString())
            box (this.Table.GetHashCode())
            box this.Name
            box (this.Organisation.GetHashCode())
            box this.Version
            HashCodes.boxHashSeq this.Authors
            HashCodes.boxHashSeq this.EndpointRepositories
            HashCodes.boxHashSeq this.Tags
            box (HashCodes.hashDateTime this.LastUpdated)
        |]
        |> HashCodes.boxHashArray 
        |> fun x -> x :?> int

File: js/Core/Template.js
import { safeHash, equals } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { Union } from "../fable_modules/fable-library-js.4.22.0/Types.js";
import { class_type, union_type, string_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";
import { defaultArg, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { newGuid } from "../fable_modules/fable-library-js.4.22.0/Guid.js";
import { ArcTable } from "./Table/ArcTable.js";
import { SemVer_tryOfString_Z721C83C5 } from "./Helper/SemVer.js";
import { ResizeArray_map } from "./Helper/Collections.js";
import { hashDateTime, boxHashSeq, boxHashArray } from "./Helper/HashCodes.js";
import { now } from "../fable_modules/fable-library-js.4.22.0/Date.js";

export class Organisation extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["DataPLANT", "Other"];
    }
    static ofString(str) {
        return (str.toLocaleLowerCase() === "dataplant") ? (new Organisation(0, [])) : (new Organisation(1, [str]));
    }
    toString() {
        const this$ = this;
        return (this$.tag === 1) ? this$.fields[0] : "DataPLANT";
    }
    IsOfficial() {
        const this$ = this;
        return equals(this$, new Organisation(0, []));
    }
}

export function Organisation_$reflection() {
    return union_type("ARCtrl.Organisation", [], Organisation, () => [[], [["Item", string_type]]]);
}

export class Template {
    constructor(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated) {
        const name_1 = defaultArg(name, "");
        const description_1 = defaultArg(description, "");
        const organisation_1 = defaultArg(organisation, new Organisation(1, ["Custom Organisation"]));
        const version_1 = defaultArg(version, "0.0.0");
        const authors_1 = defaultArg(authors, []);
        const repos_1 = defaultArg(repos, []);
        const tags_1 = defaultArg(tags, []);
        const lastUpdated_1 = defaultArg(lastUpdated, now());
        this["Id@"] = id;
        this["Table@"] = table;
        this["Name@"] = name_1;
        this["Description@"] = description_1;
        this["Organisation@"] = organisation_1;
        this["Version@"] = version_1;
        this["Authors@"] = authors_1;
        this["EndpointRepositories@"] = repos_1;
        this["Tags@"] = tags_1;
        this["LastUpdated@"] = lastUpdated_1;
    }
    get Id() {
        const __ = this;
        return __["Id@"];
    }
    set Id(v) {
        const __ = this;
        __["Id@"] = v;
    }
    get Table() {
        const __ = this;
        return __["Table@"];
    }
    set Table(v) {
        const __ = this;
        __["Table@"] = v;
    }
    get Name() {
        const __ = this;
        return __["Name@"];
    }
    set Name(v) {
        const __ = this;
        __["Name@"] = v;
    }
    get Description() {
        const __ = this;
        return __["Description@"];
    }
    set Description(v) {
        const __ = this;
        __["Description@"] = v;
    }
    get Organisation() {
        const __ = this;
        return __["Organisation@"];
    }
    set Organisation(v) {
        const __ = this;
        __["Organisation@"] = v;
    }
    get Version() {
        const __ = this;
        return __["Version@"];
    }
    set Version(v) {
        const __ = this;
        __["Version@"] = v;
    }
    get Authors() {
        const __ = this;
        return __["Authors@"];
    }
    set Authors(v) {
        const __ = this;
        __["Authors@"] = v;
    }
    get EndpointRepositories() {
        const __ = this;
        return __["EndpointRepositories@"];
    }
    set EndpointRepositories(v) {
        const __ = this;
        __["EndpointRepositories@"] = v;
    }
    get Tags() {
        const __ = this;
        return __["Tags@"];
    }
    set Tags(v) {
        const __ = this;
        __["Tags@"] = v;
    }
    get LastUpdated() {
        const __ = this;
        return __["LastUpdated@"];
    }
    set LastUpdated(v) {
        const __ = this;
        __["LastUpdated@"] = v;
    }
    static make(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated) {
        return new Template(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated);
    }
    static create(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated) {
        return new Template(id, table, unwrap(name), unwrap(description), unwrap(organisation), unwrap(version), unwrap(authors), unwrap(repos), unwrap(tags), unwrap(lastUpdated));
    }
    static init(templateName) {
        return new Template(newGuid(), ArcTable.init(templateName), templateName);
    }
    get SemVer() {
        const this$ = this;
        return unwrap(SemVer_tryOfString_Z721C83C5(this$.Version));
    }
    ReferenceEquals(other) {
        const this$ = this;
        return this$ === other;
    }
    StructurallyEquals(other) {
        const this$ = this;
        return safeHash(this$) === safeHash(other);
    }
    Copy() {
        const this$ = this;
        const nextAuthors = ResizeArray_map((c) => c.Copy(), this$.Authors);
        const nextRepos = ResizeArray_map((c_1) => c_1.Copy(), this$.EndpointRepositories);
        const nextTags = ResizeArray_map((c_2) => c_2.Copy(), this$.Tags);
        return Template.create(this$.Id, this$.Table.Copy(), this$.Name, this$.Description, this$.Organisation, this$.Version, nextAuthors, nextRepos, nextTags, this$.LastUpdated);
    }
    Equals(other) {
        let template;
        const this$ = this;
        return (other instanceof Template) && ((template = other, this$.StructurallyEquals(template)));
    }
    GetHashCode() {
        let copyOfStruct;
        const this$ = this;
        return boxHashArray([(copyOfStruct = this$.Id, copyOfStruct), safeHash(this$.Table), this$.Name, safeHash(this$.Organisation), this$.Version, boxHashSeq(this$.Authors), boxHashSeq(this$.EndpointRepositories), boxHashSeq(this$.Tags), hashDateTime(this$.LastUpdated)]) | 0;
    }
}

export function Template_$reflection() {
    return class_type("ARCtrl.Template", undefined, Template);
}

export function Template_$ctor_Z17EA765C(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated) {
    return new Template(id, table, name, description, organisation, version, authors, repos, tags, lastUpdated);
}

